Cursorposition sometimes printes in terminal after upgrading to v13.7.0

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What happened?

After upgrading to version 13.7.0 i started seeing ANSI(?) cursor position being printed to terminal at the beginging of my prompt. Its not always, see attached screenshot. I have also tested v14.0.2, but still same problem. After downgrading to v13.6.0 the problem is gone.

I'm only seeing this issue in tmux(v3.2a) using bash(v5.2.15). I dont see the issue using tmux and pwsh, or using just bash without tmux. I have tested in both Alacritty (v0.11.0) and iTerm2 (v3.4.19).


![Screenshot 2023-02-01 at 07 53 02](https://user-images.githubusercontent.com/609372/215973358-7a4dc2e6-2e45-4097-a9ee-247a3ea2a69d.png)

### Theme

https://github.com/madsaune/milbo-omp-theme/blob/main/milbo.omp.yml

### What OS are you seeing the problem on?

macOS

### Which shell are you using?

bash

### Log output

```Shell
[bash] ~                                                                      07:57:00
❯ oh-my-posh debug

Version: 14.0.2

Segments:

ConsoleTitle(false) -   0 ms -
shell(true)         -   0 ms - [bash]
git(false)          -   0 ms -
path(true)          -   0 ms -  ~
time(true)          -   0 ms -  07:57:01
executiontime(false) -   0 ms -
exit(true)          -   0 ms - ❯

Run duration: 4.115502ms

Cache path: /Users/mm/.cache/oh-my-posh

Config path: /Users/mm/code/github.com/madsaune/milbo-omp-theme/milbo.omp.yml

Logs:

[DEBUG] 07:57:01.302 shell.go:Getenv:342 → NO DATA
[TRACE] 07:57:01.302 shell.go:Getenv(XDG_CACHE_HOME) - 37.012µs
[TRACE] 07:57:01.302 shell_unix.go:CachePath() - 82.583µs
[DEBUG] 07:57:01.303 shell.go:Getenv:342 → /Users/mm/code/github.com/madsaune/milbo-omp-theme/milbo.omp.yml
[TRACE] 07:57:01.303 shell.go:Getenv(POSH_THEME) - 8.444µs
[TRACE] 07:57:01.303 shell.go:resolveConfigPath() - 14.517µs
[TRACE] 07:57:01.303 shell.go:Init() - 1.342078ms
[TRACE] 07:57:01.303 shell.go:Flags() - 80ns
[TRACE] 07:57:01.304 config.go:loadConfig() - 748.574µs
[TRACE] 07:57:01.304 shell.go:Flags() - 89ns
[DEBUG] 07:57:01.304 shell.go:Getenv:342 → NO DATA
[TRACE] 07:57:01.304 shell.go:Getenv(OMP_CACHE_DISABLED) - 3.522µs
[TRACE] 07:57:01.304 shell.go:Flags() - 72ns
[TRACE] 07:57:01.304 shell.go:GOOS() - 115ns
[DEBUG] 07:57:01.304 shell.go:Pwd:370 → NO DATA
[TRACE] 07:57:01.304 shell.go:Pwd() - 29.663µs
[TRACE] 07:57:01.304 shell.go:GOOS() - 49ns
[DEBUG] 07:57:01.304 shell.go:Shell:587 → process name: bash
[TRACE] 07:57:01.304 shell.go:Shell() - 172.979µs
[TRACE] 07:57:01.304 shell.go:Flags() - 67ns
[TRACE] 07:57:01.304 shell_unix.go:Root() - 1.035µs
[TRACE] 07:57:01.304 shell.go:Shell() - 88ns
[TRACE] 07:57:01.304 shell.go:ErrorCode() - 70ns
[DEBUG] 07:57:01.304 shell.go:Getenv:342 → NO DATA
[TRACE] 07:57:01.304 shell.go:Getenv(WSL_DISTRO_NAME) - 3.22µs
[TRACE] 07:57:01.304 shell_unix.go:IsWsl() - 9.771µs
[DEBUG] 07:57:01.304 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.304 shell.go:Pwd() - 5.614µs
[TRACE] 07:57:01.304 shell.go:GOOS() - 70ns
[TRACE] 07:57:01.304 shell.go:GOOS() - 48ns
[TRACE] 07:57:01.304 shell.go:GOOS() - 48ns
[TRACE] 07:57:01.304 shell.go:GOOS() - 48ns
[DEBUG] 07:57:01.304 shell.go:User:491 → mm
[TRACE] 07:57:01.304 shell.go:User() - 2.899µs
[DEBUG] 07:57:01.304 shell.go:Host:503 → carlin
[TRACE] 07:57:01.304 shell.go:Host() - 14.01µs
[TRACE] 07:57:01.304 shell.go:GOOS() - 48ns
[TRACE] 07:57:01.304 shell.go:TemplateCache() - 85.298µs
[TRACE] 07:57:01.304 shell.go:TemplateCache() - 205ns
[TRACE] 07:57:01.304 shell.go:Shell() - 90ns
[DEBUG] 07:57:01.304 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.304 shell.go:Pwd() - 4.618µs
[TRACE] 07:57:01.304 shell.go:GOOS() - 59ns
[DEBUG] 07:57:01.304 shell.go:Getenv:342 → NO DATA
[TRACE] 07:57:01.304 shell.go:Getenv(WSL_DISTRO_NAME) - 2.905µs
[TRACE] 07:57:01.304 shell_unix.go:IsWsl() - 11.638µs
[TRACE] 07:57:01.304 shell_unix.go:IsWsl2() - 14.698µs
[TRACE] 07:57:01.304 shell.go:GOOS() - 48ns
[DEBUG] 07:57:01.304 shell.go:CommandPath:542 → /usr/local/bin/git
[TRACE] 07:57:01.304 shell.go:CommandPath(git) - 42.286µs
[TRACE] 07:57:01.304 shell.go:HasCommand(git) - 45.686µs
[DEBUG] 07:57:01.304 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.304 shell.go:Pwd() - 2.948µs
[ERROR] 07:57:01.305 shell.go:HasParentFilePath:675 → stat /.git: no such file or directory
[TRACE] 07:57:01.305 shell.go:HasParentFilePath(.git) - 32.558µs
[DEBUG] 07:57:01.305 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.305 shell.go:Pwd() - 3.21µs
[TRACE] 07:57:01.305 shell.go:GOOS() - 50ns
[DEBUG] 07:57:01.305 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.305 shell.go:Pwd() - 3.141µs
[TRACE] 07:57:01.305 shell.go:Shell() - 99ns
[TRACE] 07:57:01.305 shell.go:Shell() - 56ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 50ns
[TRACE] 07:57:01.305 shell.go:PathSeparator() - 104ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 60ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 54ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 88ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 50ns
[TRACE] 07:57:01.305 shell.go:PathSeparator() - 50ns
[TRACE] 07:57:01.305 shell.go:PathSeparator() - 56ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 54ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 53ns
[TRACE] 07:57:01.305 shell.go:PathSeparator() - 54ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 53ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 55ns
[TRACE] 07:57:01.305 shell.go:PathSeparator() - 55ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 59ns
[TRACE] 07:57:01.305 shell.go:GOOS() - 72ns
[TRACE] 07:57:01.305 shell.go:PathSeparator() - 76ns
[DEBUG] 07:57:01.305 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.305 shell.go:Pwd() - 3.057µs
[DEBUG] 07:57:01.305 shell.go:Getenv:342 → NO DATA
[TRACE] 07:57:01.305 shell.go:Getenv(WSL_DISTRO_NAME) - 2.844µs
[TRACE] 07:57:01.305 shell_unix.go:IsWsl() - 5.921µs
[TRACE] 07:57:01.305 shell.go:StackCount() - 93ns
[TRACE] 07:57:01.305 shell_unix.go:DirIsWritable(/Users/mm) - 14.913µs
[TRACE] 07:57:01.305 shell.go:TemplateCache() - 110ns
[TRACE] 07:57:01.305 shell.go:TemplateCache() - 137ns
[TRACE] 07:57:01.305 shell.go:Shell() - 95ns
[TRACE] 07:57:01.305 shell.go:Flags() - 98ns
[DEBUG] 07:57:01.305 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.305 shell.go:Pwd() - 3.629µs
[TRACE] 07:57:01.305 shell.go:GOOS() - 51ns
[TRACE] 07:57:01.305 shell.go:TemplateCache() - 75ns
[TRACE] 07:57:01.305 shell.go:TemplateCache() - 139ns
[TRACE] 07:57:01.305 shell.go:Shell() - 85ns
[DEBUG] 07:57:01.305 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.305 shell.go:Pwd() - 5.917µs
[TRACE] 07:57:01.305 shell.go:GOOS() - 56ns
[TRACE] 07:57:01.305 shell.go:ExecutionTime() - 90ns
[TRACE] 07:57:01.305 shell.go:Flags() - 64ns
[DEBUG] 07:57:01.305 shell.go:Pwd:350 → /Users/mm
[TRACE] 07:57:01.305 shell.go:Pwd() - 3.139µs
[TRACE] 07:57:01.305 shell.go:GOOS() - 50ns
[TRACE] 07:57:01.305 shell.go:ErrorCode() - 99ns
[TRACE] 07:57:01.305 shell.go:TemplateCache() - 79ns
[TRACE] 07:57:01.305 shell.go:Shell() - 85ns
[TRACE] 07:57:01.306 shell.go:TemplateCache() - 138ns
[DEBUG] 07:57:01.306 shell.go:Getenv:342 → NO DATA
[TRACE] 07:57:01.306 shell.go:Getenv(XDG_CACHE_HOME) - 3.959µs
[TRACE] 07:57:01.306 shell_unix.go:CachePath() - 28.401µs
[TRACE] 07:57:01.306 shell.go:Flags() - 61ns
^[[24;1R[bash] ~                                                              07:57:01
❯
```

--- a/src/engine/engine.go
+++ b/src/engine/engine.go
@@ -16,11 +16,10 @@ var (
 )
 
 type Engine struct {
-	Config      *Config
-	Env         platform.Environment
-	Writer      *ansi.Writer
-	Plain       bool
-	PromptCount int
+	Config *Config
+	Env    platform.Environment
+	Writer *ansi.Writer
+	Plain  bool
 
 	console           strings.Builder
 	currentLineLength int
@@ -64,9 +63,13 @@ func (e *Engine) canWriteRightBlock(rprompt bool) bool {
 func (e *Engine) PrintPrimary() string {
 	// cache a pointer to the color cycle
 	cycle = &e.Config.Cycle
-	firstLine := e.Env.Getenv("POSH_CURSOR_LINE") == "1"
 	for i, block := range e.Config.Blocks {
-		e.renderBlock(block, (i == 0 && (e.PromptCount == 1 || firstLine)))
+		var cancelNewline bool
+		if i == 0 {
+			row, _ := e.Env.CursorPosition()
+			cancelNewline = e.Env.Flags().PromptCount == 1 || row == 1
+		}
+		e.renderBlock(block, cancelNewline)
 	}
 	if len(e.Config.ConsoleTitleTemplate) > 0 {
 		title := e.getTitleTemplateText()

--- a/src/engine/new.go
+++ b/src/engine/new.go
@@ -25,11 +25,10 @@ func New(flags *platform.Flags) *Engine {
 	ansiWriter.Init(env.Shell())
 
 	eng := &Engine{
-		Config:      cfg,
-		Env:         env,
-		Writer:      ansiWriter,
-		Plain:       flags.Plain,
-		PromptCount: env.CmdFlags.PromptCount,
+		Config: cfg,
+		Env:    env,
+		Writer: ansiWriter,
+		Plain:  flags.Plain,
 	}
 
 	return eng

--- a/src/mock/environment.go
+++ b/src/mock/environment.go
@@ -268,3 +268,8 @@ func (env *MockedEnvironment) DirIsWritable(path string) bool {
 func (env *MockedEnvironment) SetPromptCount() {
 	_ = env.Called()
 }
+
+func (env *MockedEnvironment) CursorPosition() (int, int) {
+	args := env.Called()
+	return args.Int(0), args.Int(1)
+}

--- a/src/platform/shell.go
+++ b/src/platform/shell.go
@@ -216,6 +216,7 @@ type Environment interface {
 	TemplateCache() *TemplateCache
 	LoadTemplateCache()
 	SetPromptCount()
+	CursorPosition() (row, col int)
 	Debug(message string)
 	Error(err error)
 	Trace(start time.Time, args ...string)
@@ -829,6 +830,16 @@ func (env *Shell) SetPromptCount() {
 		env.Cache().Set(PROMPTCOUNTCACHE, strconv.Itoa(count), 1440)
 	}
 	env.CmdFlags.PromptCount = count
+}
+
+func (env *Shell) CursorPosition() (row, col int) {
+	if number, err := strconv.Atoi(env.Getenv("POSH_CURSOR_LINE")); err == nil {
+		row = number
+	}
+	if number, err := strconv.Atoi(env.Getenv("POSH_CURSOR_COLUMN")); err != nil {
+		col = number
+	}
+	return
 }
 
 func IsPathSeparator(env Environment, c uint8) bool {

--- a/src/shell/scripts/omp.bash
+++ b/src/shell/scripts/omp.bash
@@ -9,20 +9,6 @@ PS0='${omp_start_time:0:$((omp_start_time="$(_omp_start_timer)",0))}'
 # set secondary prompt
 PS2="$(::OMP:: print secondary --config="$POSH_THEME" --shell=bash --shell-version="$BASH_VERSION")"
 
-function _set_posh_cursor_position() {
-    # not supported in Midnight Commander
-    # see https://github.com/JanDeDobbeleer/oh-my-posh/issues/3415
-    if [[ -v MC_SID ]] || [[ $TERM_PROGRAM == "WarpTerminal" ]]; then
-        return
-    fi
-    local pos
-    echo -ne "\033[6n"            # ask the terminal for the position
-    read -s -d\[ garbage          # discard the first part of the response
-    read -s -d R pos              # store the position in bash variable 'pos'
-    export POSH_CURSOR_LINE=${pos%;*}
-    export POSH_CURSOR_COLUMN=${pos#*;}
-}
-
 function _omp_start_timer() {
     ::OMP:: get millis
 }
@@ -42,7 +28,6 @@ function _omp_hook() {
         omp_start_time=""
     fi
     set_poshcontext
-    _set_posh_cursor_position
     PS1="$(::OMP:: print primary --config="$POSH_THEME" --shell=bash --shell-version="$BASH_VERSION" --error="$ret" --execution-time="$omp_elapsed" --stack-count="$omp_stack_count" | tr -d '\0')"
     return $ret
 }

--- a/src/shell/scripts/omp.zsh
+++ b/src/shell/scripts/omp.zsh
@@ -7,20 +7,6 @@ export POSH_PROMPT_COUNT=0
 # set secondary prompt
 PS2="$(::OMP:: print secondary --config="$POSH_THEME" --shell=zsh)"
 
-function _set_posh_cursor_position() {
-  # not supported in Midnight Commander
-  # see https://github.com/JanDeDobbeleer/oh-my-posh/issues/3415
-  if [[ -v MC_SID ]] || [[ $TERM_PROGRAM == "WarpTerminal" ]]; then
-      return
-  fi
-  local pos
-  echo -ne "\033[6n"            # ask the terminal for the position
-  read -s -d\[ garbage          # discard the first part of the response
-  read -s -d R pos              # store the position in bash variable 'pos'
-  export POSH_CURSOR_LINE=${pos%;*}
-  export POSH_CURSOR_COLUMN=${pos#*;}
-}
-
 # template function for context loading
 function set_poshcontext() {
   return
@@ -41,7 +27,6 @@ function prompt_ohmyposh_precmd() {
   count=$((POSH_PROMPT_COUNT+1))
   export POSH_PROMPT_COUNT=$count
   set_poshcontext
-  _set_posh_cursor_position
   eval "$(::OMP:: print primary --config="$POSH_THEME" --error="$omp_last_error" --execution-time="$omp_elapsed" --stack-count="$omp_stack_count" --eval --shell=zsh --shell-version="$ZSH_VERSION")"
   unset omp_start_time
 }

--- a/website/docs/configuration/block.mdx
+++ b/website/docs/configuration/block.mdx
@@ -86,9 +86,8 @@ Tells the engine what to do with the block. There are two options:
 
 ### Newline
 
-Start the block on a new line - defaults to `false`. For `zsh`, `bash`, `pwsh` this will not print a newline that's
-defined on the first block when the prompt is on the first line (when using clear), or when the shell session starts
-(1st prompt).
+Start the block on a new line - defaults to `false`. For `pwsh` and `cmd` this will not print a newline that's defined
+on the first block when the prompt is on the first line (when using clear), or when the shell session starts (1st prompt).
 
 ### Alignment
 

